### PR TITLE
Add `trim` to remove trailing whitespace

### LINKED ISSUE #1481: Add `trim_file` command and `trim_file_on_save` configuration option
#### Describe your feature request

I know there is an `trim_selection` command but using it breaks the current selection, which is a not very helpful. It's also not very practical to do it for each opened file when ending a somewhat long session.

### PR DESCRIPTION
**Pull request by kirawi**
Resolves #1481

~Currently it doesn't delete empty trailing lines. Is this a desired behavior?~

### CODE REVIEW

File: helix-term/src/commands/typed.rs [hunk 1]
```diff
@@ -1515,6 +1515,28 @@ fn run_shell_command(
     Ok(())
 }
 
+fn trim_whitespace(
+    cx: &mut compositor::Context,
+    _: &[Cow<str>],
+    _: PromptEvent,
+) -> anyhow::Result<()> {
+    let mut pos = 0;
```
Comment (kirawi) [inline helix-term/src/commands/typed.rs:1523]: I wonder if it would be useful to have a `Transaction::change_by_line`.

**Comment by poliorcetics** [comment]
> Currently it doesn't delete empty trailing lines. Is this a desired behavior?

For me yes, but i don't know if others agree

File: helix-term/src/commands/typed.rs [hunk 2]
```diff
@@ -1515,6 +1515,64 @@ fn run_shell_command(
...
+                } else {
+                    pos -= line.len_chars();
+                    None
+                }
+            })
+            .unwrap_or(0);
+        let end = doc.text().len_chars();
+        let transaction = Transaction::change(doc.text(), [(start, end, None)].into_iter());
+        doc.apply(&transaction, view.id);
+    }
```
[the remark this answers is not in the document]
Comment (kirawi) [reply helix-term/src/commands/typed.rs:1570]: It trims all the trailing lines in the file. Should I delete this?

**Comment by SoraTenshi** [comment]
i am not sure if this may be desired, but an option to trim on `:write` would be decent i guess?
(obv. also `:write-quit`)

**Comment by kirawi** [comment]
> i am not sure if this may be desired, but an option to trim on `:write` would be decent i guess? (obv. also `:write-quit`)

Someone could follow up on that with a PR.

**Comment by SoraTenshi** [comment]
> > i am not sure if this may be desired, but an option to trim on `:write` would be decent i guess? (obv. also `:write-quit`)
> 
> Someone could follow up on that with a PR. I'm not personally interested in implementing it.

I guess this gonna be me then. :)
Luckily i get a notification once this PR is merged.

**Comment by koalalorenzo** [comment]
Is this still being worked on? 👀

**Comment by rcorre** [comment]
I'm unsure @koalalorenzo, but as a workaround you can use `git-stripspace` as a formatter: https://github.com/helix-editor/helix/issues/1481#issuecomment-1421640827

```formatter = { command = "git", args = ["stripspace"] }```

That will override an existing formatter, but I figure most language formatters (`clang-format`, `gofmt`, `rustfmt`, etc.) will already strip space, so this is mainly needed for filetypes that don't have formatters.

**Comment by Joe-Zer0** [comment]
I would eagerly looking forward to a command to trim trailing whitespace (and eventually trim on save as well).  Any update on when this will be merged?  Thanks!

**Comment by tshepang** [comment]
oh, didn't think of that (I suppose via `:pipe cmd`?)... what cmd do you use

**Comment by tgross35** [comment]
The "trim final newlines" (`:trim lines`) option of this somewhat relates to https://github.com/helix-editor/helix/pull/8157 which inserts a final newline if there is none. That is enabled on `:w` but not `:wa`, which matches vim.

If this PR gets refactored, enabling "trim trailing whitespace from all lines" (`:trim spaces`) on `:w` would likely be a good idea, this would also match vim. `:trim lines` shouldn't be on by default.

`:trim spaces` doesn't seem like the most accurate command name, since it doesn't just remove any spaces. Not sure what would be better though - maybe `:trim trailing-whitespace` with an alias `:trim trailing`?

**Comment by kirawi** [comment]
Superseded by #8366

**Pull request closed without merging**